File: autoresearch/lib/run_lane.py
#!/usr/bin/env python3
"""HP-1 autoresearch lane scorer. Produces autoresearch/lane-<x>/score.json +
appends the append-only ledger. REUSES the live HP-1 engine (engine/hp1_engine.py,
the corrected t+1 simulator) as the held-out scorer — it never reimplements the
factor math (AUTORESEARCH_DOCTRINE.md Part I).

Usage (from repo root, in the HP-1 venv):
  python autoresearch/lib/run_lane.py <A|B|C> [--label <tag>]

HARD GATE (SEC 206(4)-1): every number here is a harness/test metric on the
corrected backtest record. It is NOT live performance and never ships as a track
record. Per HP-1's own red-team (docs/hp1/HP1_redteam_findings.md), the published
110.5%/2.43 record was inflated by a now-fixed execution lookahead, and the
risk-adjusted edge over equal-weighting the same universe is ~0. Lane B's HEADLINE
is therefore the HONEST EDGE (vs equal-weight, with a bootstrap CI), NOT raw Sharpe.
"""
import sys, os, json, argparse, subprocess
from datetime import datetime, timezone

REPO = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", ".."))
ENGINE = os.path.join(REPO, "engine")
sys.path.insert(0, ENGINE)
sys.path.insert(0, os.path.dirname(__file__))


def git_sha():
    try:
        return subprocess.check_output(["git", "rev-parse", "HEAD"], cwd=REPO).decode().strip()
    except Exception:
        return "UNKNOWN"


def stamp():
    return os.environ.get("SCORE_TS") or datetime.now(timezone.utc).isoformat()


def run_pytest(paths, cwd=None):
    """Run pytest on paths; return {passed,total,green} parsed from the summary.
    cwd defaults to the repo root; engine tests must run from engine/ so the
    `import hp1_engine` resolves."""
    try:
        out = subprocess.run(
            [sys.executable, "-m", "pytest", "-q", *paths],
            cwd=cwd or REPO, capture_output=True, text=True, timeout=300,
        )
        text = out.stdout + out.stderr
    except Exception as e:
        return {"passed": None, "total": None, "green": False, "error": str(e)}
    import re
    passed = int(m.group(1)) if (m := re.search(r"(\d+) passed", text)) else 0
    failed = int(m.group(1)) if (m := re.search(r"(\d+) failed", text)) else 0
    return {"passed": passed, "failed": failed, "total": passed + failed,
            "green": out.returncode == 0 and failed == 0 and passed > 0}


# ── LANE A — engine determinism + factor invariants (offline, synthetic) ──────
def lane_a():
    import numpy as np, pandas as pd
    import hp1_engine as eng

    # Deterministic synthetic panel (fixed seed). No network.
    idx = pd.bdate_range("2024-01-01", periods=340)
    n = len(idx)
    rng = np.random.default_rng(7)
    cols = {}
    cols["UP"] = pd.Series(np.linspace(100, 220, n), idx)            # steady riser
    cols["DN"] = pd.Series(np.linspace(100, 52, n), idx)             # steady decliner
    cols["CH"] = pd.Series(100 * np.cumprod(1 + rng.normal(0, .04, n)), idx)  # chop
    for k in range(9):  # extra names so z-scores have a real cross-section
        drift = rng.uniform(-0.0004, 0.0009)
        cols[f"N{k}"] = pd.Series(100 * np.cumprod(1 + rng.normal(drift, .02, n)), idx)
    px = pd.DataFrame(cols)

    f1 = eng.factors(px, idx[-1])
    f2 = eng.factors(px, idx[-1])
    deterministic = f1.equals(f2)

    inv = {
        "riser_outranks_decliner": bool(f1.loc["UP", "tac"] > f1.loc["DN", "tac"]),
        "trend_gate_riser_above100": bool(f1.loc["UP", "above100"]),
        "trend_gate_decliner_below100": bool(not f1.loc["DN", "above100"]),
        "smooth_riser_beats_chop_on_RAM": bool(f1.loc["UP", "zRAM"] > f1.loc["CH", "zRAM"]),
        "zM_finite": bool(np.isfinite(f1["zM"].dropna()).all()),
    }
    # weights single-name cap holds on a feasible (>=7-name) book
    sel = f1[f1["above100"]].sort_values("tac", ascending=False).head(10).index.tolist()
    wcap_ok = True
    if len(sel) >= 7:
        w = eng.weights(sel, f1)
        wcap_ok = bool(w.max() <= 0.15 + 1e-6 and abs(w.sum() - 1) < 1e-9)
    inv["weights_cap_<=0.15_on_feasible_book"] = wcap_ok

    reg = run_pytest(["tests/"], cwd=ENGINE)
    invariants_pass = all(inv.values())
    return {
        "metric": {
            "name": "engine_determinism_and_factor_invariants",
            "deterministic": deterministic,
            "invariants": inv,
            "invariants_pass": invariants_pass,
            "universe_n": len(eng.UNIV),
        },
        "regression": reg,
        "runnable_offline": True,
        "pass": bool(deterministic and invariants_pass and reg["green"]),
    }


# ── LANE B — reproduce corrected v2 record + HONEST edge (bootstrap CI) ────────
def _sharpe(daily, rf=0.04):
    import numpy as np
    if len(daily) < 2:
        return 0.0
    ann_ret = (1 + daily).prod() ** (252 / len(daily)) - 1
    vol = daily.std() * np.sqrt(252)
    return float((ann_ret - rf) / vol) if vol else 0.0


def lane_b(windows=None):
    import numpy as np, pandas as pd, yfinance as yf
    import hp1_engine as eng
    from restate_record import IPO_COHORT

    windows = windows or [("24M", "2024-06-10"), ("36M", "2023-06-12")]
    end = "2026-06-10"

    raw = yf.download(eng.UNIV + eng.BENCH, start="2022-06-01", end=None,
                      auto_adjust=True, progress=False)
    px = raw["Close"].dropna(axis=1, how="all")
    rets = px.pct_change()
    upx = px[[c for c in eng.UNIV if c in px.columns]]
    urets = rets[[c for c in eng.UNIV if c in px.columns]]
    names_loaded = int(upx.shape[1])

    # committed corrected-record tie-out targets
    committed = {}
    for label in ("24M", "36M"):
        p = os.path.join(ENGINE, "data", f"results_{label.lower()}_v2.csv")
        if os.path.exists(p):
            committed[label] = pd.read_csv(p).set_index("name")

    rng = np.random.default_rng(0)  # deterministic bootstrap

    per_window = {}
    for label, start in windows:
        v2, _ = eng.simulate(upx, urets, start, end, "tac", 10, gate=True)
        ew = (1 + urets.loc[start:end].mean(axis=1).fillna(0)).cumprod()
        ex = [c for c in upx.columns if c not in IPO_COHORT]
        v2x, _ = eng.simulate(upx[ex], urets[ex], start, end, "tac", 10, gate=True)
        ewx = (1 + urets[ex].loc[start:end].mean(axis=1).fillna(0)).cumprod()

        s_v2, s_ew = _sharpe(v2.pct_change().dropna()), _sharpe(ew.pct_change().dropna())
        s_v2x, s_ewx = _sharpe(v2x.pct_change().dropna()), _sharpe(ewx.pct_change().dropna())

        # tie-out vs committed corrected record
        tie = None
        if label in committed and "V2 Tactical+gate" in committed[label].index:
            s_committed = float(committed[label].loc["V2 Tactical+gate", "sharpe"])
            denom = abs(s_committed) if s_committed else 1.0
            tie = {
                "v2_sharpe_live": round(s_v2, 3),
                "v2_sharpe_committed": round(s_committed, 3),
                "pct_diff": round(abs(s_v2 - s_committed) / denom, 4),
                "within_10pct": abs(s_v2 - s_committed) / denom <= 0.10,
            }

        # HONEST edge: bootstrap CI on the Sharpe-delta of (V2 - EW) daily returns.
        d = (v2.pct_change() - ew.pct_change()).dropna().values
        boot = []
        for _ in range(2000):
            samp = rng.choice(d, size=len(d), replace=True)
            mu, sd = samp.mean(), samp.std()
            boot.append((mu * 252) / (sd * np.sqrt(252)) if sd else 0.0)
        boot.sort()
        ci_lo, ci_hi = float(boot[49]), float(boot[1949])  # 2.5% / 97.5%

        per_window[label] = {
            "v2_sharpe": round(s_v2, 3),
            "ew50_sharpe": round(s_ew, 3),
            "edge_vs_ew_sharpe": round(s_v2 - s_ew, 3),
            "edge_vs_ew_ex_ipo_sharpe": round(s_v2x - s_ewx, 3),
            "edge_bootstrap_ci95": [round(ci_lo, 3), round(ci_hi, 3)],
            "edge_distinguishable_from_zero": bool(ci_lo > 0 or ci_hi < 0),
            "tie_out": tie,
        }

    tie_ok = all(w["tie_out"] and w["tie_out"]["within_10pct"] for w in per_window.values() if w["tie_out"])
    return {
        "metric": {
            "name": "v2_record_reproduction_and_honest_edge",
            "anchor": "HP-1's own corrected (t+1) record, engine/data/results_*_v2.csv (V2 Tactical+gate). The headline is the HONEST EDGE vs equal-weight, not raw Sharpe.",
            "names_loaded": names_loaded,
            "windows": per_window,
            "redteam_note": "Per docs/hp1/HP1_redteam_findings.md: honest risk-adjusted edge over equal-weighting the same universe is ~0 (+0.05 24m / -0.01 36m ex-IPO). If the bootstrap CI straddles 0, the edge is statistically indistinguishable from zero — that is the expected, honest result, not a failure.",
        },
        "regression": {"note": "tie-out IS the regression for Lane B"},
        "runnable_offline": False,
        "tie_out_within_10pct": tie_ok,
        "pass": None,  # Lane B's job is honest measurement, not a pass/fail target
    }


# ── LANE C — Fable review citation validation ─────────────────────────────────
def lane_c():
    from fable_citations import validate_fable_review
    CITE = {"claim": "guidance cut", "source": "Reuters", "date": "2026-06-20", "url": "https://r.com/x"}
    CITE2 = {"claim": "contract", "source": "8-K", "date": "2026-06-21", "url": "https://sec.gov/y"}
    clean = [
        {"ticker": "NVDA", "verdict": "CONFIRM", "adjustment_pct": 0, "evidence": []},
        {"ticker": "TSM", "verdict": "DOWNGRADE", "adjustment_pct": -10, "confidence": "H", "evidence": [CITE]},
        {"ticker": "VRT", "verdict": "UPGRADE", "adjustment_pct": 5, "confidence": "H", "evidence": [CITE, CITE2]},
    ]
    audit = validate_fable_review(clean)
    reg = run_pytest(["autoresearch/lib/fable_citations_test.py"])
    return {
        "metric": {
            "name": "fable_review_citation_leak_rate",
            "validator": "autoresearch/lib/fable_citations.py (validate_fable_review)",
            "contract": "FABLE_REVIEW_RUBRIC_v1.md §5/§6/§50: every DOWNGRADE/UPGRADE/VETO must carry the required dated {source,date,url} citations; uncited adjustments demote to FLAG. L-confidence findings cannot exceed -5.",
            "clean_fixture_leak_rate": audit.leak_rate,
            "clean_fixture_unhandled": audit.unhandled,
            "status": "VALIDATOR_BUILT_LIVE_PENDING",
            "live_pending": "URL-liveness (§50 server-side fetch: page resolves + contains the claimed entity/date) + wiring as the Fable orchestrator's post-output gate. Offline half (count/shape/date/magnitude) is enforced + unit-green here.",
        },
        "regression": reg,
        "runnable_offline": True,
        "pass": bool(audit.leak_rate == 0.0 and audit.unhandled == 0 and reg["green"]),
    }


LANES = {
    "A": ("Engine determinism + factor invariants (50-name HP-1 universe)", lane_a),
    "B": ("Reproduce corrected v2 record + honest edge vs equal-weight (bootstrap CI)", lane_b),
    "C": ("Fable review citation-validation leak rate -> 0 (rubric §6/§50)", lane_c),
}


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("lane", choices=["A", "B", "C"])
    ap.add_argument("--label", default="run")
    args = ap.parse_args()

    title, fn = LANES[args.lane]
    body = fn()
    score = {
        "schema": "autoresearch/score.v2-hp1",
        "engine": "HP-1 (engine/hp1_engine.py — corrected t+1 simulator)",
        "lane": args.lane,
        "title": title,
        "label": args.label,
        "commit": git_sha(),
        "generated_at": stamp(),
        "python": sys.version.split()[0],
        **body,
    }
    out_dir = os.path.join(REPO, "autoresearch", f"lane-{args.lane.lower()}")
    os.makedirs(out_dir, exist_ok=True)
    with open(os.path.join(out_dir, "score.json"), "w") as fh:
        json.dump(score, fh, indent=2)
        fh.write("\n")

    # append-only ledger
    if args.lane == "A":
        head = {"deterministic": body["metric"]["deterministic"],
                "invariants_pass": body["metric"]["invariants_pass"],
                "tests": body["regression"].get("passed")}
    elif args.lane == "B":
        head = {"windows": {k: {"edge_vs_ew": v["edge_vs_ew_sharpe"],
                                "edge_ci95": v["edge_bootstrap_ci95"],
                                "distinguishable": v["edge_distinguishable_from_zero"],
                                "tie_out_10pct": (v["tie_out"] or {}).get("within_10pct")}
                            for k, v in body["metric"]["windows"].items()}}
    else:
        head = {"clean_leak_rate": body["metric"]["clean_fixture_leak_rate"],
                "status": body["metric"]["status"], "tests": body["regression"].get("passed")}
    row = {"ts": score["generated_at"], "lane": args.lane, "label": args.label,
           "commit": score["commit"], "pass": score["pass"], "headline": head}
    with open(os.path.join(REPO, "autoresearch", "score_ledger.jsonl"), "a") as fh:
        fh.write(json.dumps(row) + "\n")

    print(f"lane {args.lane} ({args.label}) -> autoresearch/lane-{args.lane.lower()}/score.json")
    print(json.dumps(head, indent=2))


if __name__ == "__main__":
    main()

File: autoresearch/lib/run_lane.mjs
#!/usr/bin/env node
// Autoresearch lane scorer — produces autoresearch/lane-<x>/score.json + appends
// the append-only ledger. Zero-dependency Node (>=23.6 / v24), runs the EXISTING
// deterministic harnesses (node:test) and the live scoring module — it never
// reimplements engine math, it reuses it (per AUTORESEARCH_DOCTRINE.md Part I).
//
// Usage:
//   node --experimental-strip-types autoresearch/lib/run_lane.mjs <A|B|C> [--label baseline]
//
// Each lane writes a versioned score.json (the "investor artifact" of
// AUTORESEARCH_DOCTRINE.md Part VI) and appends one row to
// autoresearch/score_ledger.jsonl. The --label tags the run (e.g. "baseline").
//
// HARD GATE (SEC 206(4)-1): the numbers here are HARNESS/TEST metrics on fixed
// fixtures and a hand-scored slate. They are NOT live performance and must never
// be presented as a track record. See program.md §Hard gates.

import { execFileSync } from "node:child_process";
import { readFileSync, writeFileSync, appendFileSync, mkdirSync } from "node:fs";
import { dirname, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";

const __dirname = dirname(fileURLToPath(import.meta.url));
const REPO = resolve(__dirname, "..", "..");
const SHARED = join(REPO, "supabase", "functions", "_shared");

// ── lane wiring ──────────────────────────────────────────────────────────────
// Each lane names: the deterministic regression files it runs, plus a headline
// metric. Files are relative to supabase/functions/_shared.
const LANES = {
  A: {
    id: "A",
    title: "Engine determinism + 20-name hand-scored slate (±5)",
    surface:
      "supabase/functions/_shared/factor-*.ts, composite.ts, concentration.ts, weekly-ranking.ts, stats.ts, metrics.ts",
    regression: [
      "factor-q.test.ts", "factor-g.test.ts", "factor-v.test.ts",
      "factor-m.test.ts", "factor-s.test.ts", "factor-insider.test.ts",
      "composite.test.ts", "concentration.test.ts", "weekly-ranking.test.ts",
      "stats.test.ts", "metrics.test.ts",
    ],
  },
  B: {
    id: "B",
    title: "Composite tie-out + THS-64 walk-forward backtest (v2 ±10% Sharpe)",
    surface:
      "supabase/functions/_shared/composite.ts (tie-out), backtest.ts (THS-64, reused not rebuilt)",
    regression: ["composite.test.ts", "backtest.test.ts"],
  },
  C: {
    id: "C",
    title: "Memo citation-validation leak rate → 0 (held-out tickers)",
    surface:
      "supabase/functions/_shared/aiq-drafts.ts (parse/range validation), memo-context.ts",
    regression: ["aiq-drafts.test.ts", "memo-context.test.ts"],
  },
};

// ── helpers ──────────────────────────────────────────────────────────────────
function gitSha() {
  try {
    return execFileSync("git", ["rev-parse", "HEAD"], { cwd: REPO })
      .toString().trim();
  } catch {
    return "UNKNOWN";
  }
}

function isoStamp() {
  // Deterministic-friendly: callers may override via SCORE_TS for reproducible
  // ledger rows; otherwise wall clock.
  return process.env.SCORE_TS || new Date().toISOString();
}

// Run node:test over a set of files with the TAP reporter and parse the summary.
function runTests(files) {
  const args = [
    "--test", "--experimental-strip-types", "--test-reporter=tap",
    ...files.map((f) => join(SHARED, f)),
  ];
  let out = "";
  let ok = true;
  try {
    out = execFileSync(process.execPath, args, { cwd: REPO }).toString();
  } catch (e) {
    // node --test exits non-zero on any failing test; capture its stdout anyway.
    out = (e.stdout ? e.stdout.toString() : "") + (e.stderr ? e.stderr.toString() : "");
    ok = false;
  }
  const grab = (re) => {
    const m = out.match(re);
    return m ? Number(m[1]) : null;
  };
  const passed = grab(/^# pass (\d+)/m);
  const failed = grab(/^# fail (\d+)/m);
  const total = grab(/^# tests (\d+)/m);
  return {
    files,
    tests: total,
    passed,
    failed,
    all_green: ok && failed === 0 && (passed ?? 0) > 0,
  };
}

// ── Lane A headline: 20-name slate tie-out (reuses computeComposite) ──────────
async function slateTieOut() {
  const { computeComposite } = await import(join(SHARED, "composite.ts"));
  const slate = JSON.parse(
    readFileSync(join(REPO, "autoresearch", "lane-a", "slate.json"), "utf-8"),
  );
  const tol = slate.tolerance ?? 5;
  const nullGauges = { naaim: null, aaii_3wk_spread: null, fear_greed: null };

  const compute = () =>
    slate.names.map((n) => {
      const r = computeComposite(
        n.ticker, n.layer,
        { q: n.q, g: n.g, v: n.v, aiq: n.aiq, m: null, s: null },
        nullGauges, n.conc_tax,
      );
      const got = r.compositeTaxed;
      const delta = got == null ? null : Number((got - n.expected_final).toFixed(4));
      return {
        ticker: n.ticker,
        expected_final: n.expected_final,
        computed: got == null ? null : Number(got.toFixed(4)),
        delta,
        within_tol: delta != null && Math.abs(delta) <= tol,
      };
    });

  // Determinism: compute twice, require byte-identical JSON.
  const run1 = compute();
  const run2 = compute();
  const deterministic = JSON.stringify(run1) === JSON.stringify(run2);

  const within = run1.filter((x) => x.within_tol).length;
  const deltas = run1.filter((x) => x.delta != null).map((x) => Math.abs(x.delta));
  return {
    tolerance: tol,
    names_total: slate.names.length,
    names_within_tol: within,
    max_abs_delta: deltas.length ? Number(Math.max(...deltas).toFixed(4)) : null,
    mean_abs_delta: deltas.length
      ? Number((deltas.reduce((a, b) => a + b, 0) / deltas.length).toFixed(4))
      : null,
    deterministic,
    per_name: run1,
    pass: deterministic && within === slate.names.length,
    source: slate._source,
  };
}

// ── per-lane builders ─────────────────────────────────────────────────────────
async function buildLaneA() {
  const reg = runTests(LANES.A.regression);
  const slate = await slateTieOut();
  return {
    metric: {
      name: "slate_tie_out_within_5",
      ...slate,
    },
    regression: reg,
    runnable_offline: true,
    pass: reg.all_green && slate.pass,
  };
}

function buildLaneB() {
  // composite tie-out + THS-64 engine invariants both run offline.
  const reg = runTests(LANES.B.regression);
  return {
    metric: {
      name: "v2_sharpe_replication_within_10pct",
      runnable_offline: false,
      status: "BLOCKED_NEEDS_HISTORY",
      reason:
        "THS-64 backtest engine is unit-green offline, but replicating the v2 hand-scored Sharpe within ±10% requires multi-year point-in-time price/score history (prices_raw + scores_history) that is not present offline. Documented in docs/SESSION_NOTES.md (THS-64 row). The v2 reference Sharpe + which v2 strategy row to anchor against is a [TERRY] methodology slot — see program.md Lane B.",
      terry_slot: true,
      reference_csv: "engine/data/results_36m_v2.csv (candidate v2 anchor — NOT yet confirmed by Terry)",
    },
    regression: reg,
    runnable_offline: false,
    pass: null, // headline cannot be asserted offline; regression is the only offline signal
  };
}

function buildLaneC() {
  // Deterministic citation/parse-validation surface runs offline; the memo
  // citation-VALIDATOR (THS-10/THS-113) that the leak-rate metric scores does
  // not exist in-tree yet — flagged honestly, not stubbed.
  const reg = runTests(LANES.C.regression);
  return {
    metric: {
      name: "memo_citation_leak_rate",
      runnable_offline: false,
      status: "NEEDS_VALIDATOR",
      reason:
        "compute-daily-memo (THS-65) synthesizes memos via Claude but ships NO deterministic citation validator in-tree as of this baseline. The nearest existing deterministic citation surface is aiq-drafts.ts (parseAiqDraft + range validation), exercised by the regression below. A memo citation-validation harness over a held-out ticker set (leak rate → 0) is the BUILD TARGET of this lane, not an artifact to reuse. THS-10/THS-113 are Linear refs, not code found in this tree this session.",
      terry_slot: true,
      held_out_set: "[TERRY] held-out ticker set not yet defined — see program.md Lane C",
    },
    regression: reg,
    runnable_offline: false,
    pass: null,
  };
}

// ── main ─────────────────────────────────────────────────────────────────────
async function main() {
  const lane = (process.argv[2] || "").toUpperCase();
  const labelIdx = process.argv.indexOf("--label");
  const label = labelIdx >= 0 ? process.argv[labelIdx + 1] : "run";
  if (!LANES[lane]) {
    console.error(`usage: run_lane.mjs <A|B|C> [--label <tag>]  (got "${process.argv[2]}")`);
    process.exit(2);
  }

  const cfg = LANES[lane];
  let body;
  if (lane === "A") body = await buildLaneA();
  else if (lane === "B") body = buildLaneB();
  else body = buildLaneC();

  const score = {
    schema: "autoresearch/score.v1",
    lane: cfg.id,
    title: cfg.title,
    label,
    editable_surface: cfg.surface,
    commit: gitSha(),
    generated_at: isoStamp(),
    node_version: process.version,
    ...body,
  };

  const outDir = join(REPO, "autoresearch", `lane-${lane.toLowerCase()}`);
  mkdirSync(outDir, { recursive: true });
  const outPath = join(outDir, "score.json");
  writeFileSync(outPath, JSON.stringify(score, null, 2) + "\n");

  // Append-only ledger (mirrors docs/EVIDENCE.md pattern, Doctrine Part VI).
  const ledgerRow = {
    ts: score.generated_at,
    lane: score.lane,
    label: score.label,
    commit: score.commit,
    pass: score.pass,
    headline:
      lane === "A"
        ? { within: body.metric.names_within_tol, of: body.metric.names_total, max_abs_delta: body.metric.max_abs_delta, deterministic: body.metric.deterministic }
        : { status: body.metric.status, regression_green: body.regression.all_green, tests: body.regression.passed },
  };
  appendFileSync(
    join(REPO, "autoresearch", "score_ledger.jsonl"),
    JSON.stringify(ledgerRow) + "\n",
  );

  console.log(`lane ${lane} (${label}) → ${outPath}`);
  console.log(JSON.stringify(ledgerRow.headline, null, 2));
}

main().catch((e) => {
  console.error("run_lane failed:", e);
  process.exit(1);
});
